src/main/java/io/cryostat/recordings/ArchiveRequestGenerator.java

```diff
@@ -0,0 +1,191 @@
+                                new Notification(
+                                        ARCHIVE_RECORDING_FAIL, Map.of("jobId", request.getId())));
+                        throw new CompletionException(e);
+                    }
+                });
+    }
+
+    @ConsumeEvent(value = ARCHIVE_ADDRESS)
+    public void onMessage(ArchiveRequest request) {
+        try {
+            performArchive(request);
+        } catch (Exception e) {
+            logger.warn("Exception thrown while servicing request: ", e);
+        }
+    }
+
+    @ConsumeEvent(value = GRAFANA_ARCHIVE_ADDRESS)
+    public void onMessage(GrafanaArchiveUploadRequest request) {
+        try {
+            logger.info("Job ID: " + request.getId() + " submitted.");
+            String result =
+                    recordingHelper
+                            .uploadToJFRDatasource(request.getPair())
+                            .await()
+                            .atMost(timeout);
+            logger.info("Grafana upload complete, firing notification");
+            bus.publish(
+                    MessagingServer.class.getName(),
+                    new Notification(GRAFANA_UPLOAD_SUCCESS, Map.of("jobId", request.getId())));
+        } catch (Exception e) {
+            logger.warn("Exception thrown while servicing request: ", e);
+            bus.publish(
+                    MessagingServer.class.getName(),
+                    new Notification(GRAFANA_UPLOAD_FAIL, Map.of("jobId", request.getId())));
+        }
+    }
+
+    @ConsumeEvent(value = GRAFANA_ACTIVE_ADDRESS)
+    public void onMessage(GrafanaActiveUploadRequest request) {
+        try {
+            logger.info("Job ID: " + request.getId() + " submitted.");
+            String result =
+                    recordingHelper
+                            .uploadToJFRDatasource(request.getTargetId(), request.getRemoteId())
+                            .await()
+                            .atMost(timeout);
+            logger.info("Grafana upload complete, firing notification");
+            bus.publish(
+                    MessagingServer.class.getName(),
+                    new Notification(GRAFANA_UPLOAD_SUCCESS, Map.of("jobId", request.getId())));
+        } catch (Exception e) {
+            logger.warn("Exception thrown while servicing request: ", e);
+            bus.publish(
+                    MessagingServer.class.getName(),
+                    new Notification(GRAFANA_UPLOAD_FAIL, Map.of("jobId", request.getId())));
+        }
+    }
+
+    public record ArchiveRequest(String id, ActiveRecording recording) {
+
+        public ArchiveRequest {
+            Objects.requireNonNull(id);
+            Objects.requireNonNull(recording);
+        }
+
+        public String getId() {
+            return id;
+        }
+
+        public ActiveRecording getRecording() {
+            return recording;
+        }
+    }
+
+    public record GrafanaArchiveUploadRequest(String id, Pair<String, String> pair) {
```

Comment:
Why is the signature for this one `String, Pair<String, String>` rather than `String, String, String`?

Reply:
If it's just because that's what the internal implementation in `ArchivedRecordings` happens to have available, I'd still rather have this event structure that with separately named fields so that the request event has a more meaningful structure to it.

Reply:
The underlying recordinghelper call it was built around takes a pair that the ArchivedRecordings handler builds from storage, I just preserved the structure that was in place.